File: frontend/src/features/conversation/contextStatus.ts
import type { ContextStatusRead } from "../../api/conversationTypes";

/** 将最小上下文元数据转为提示，读取范围不表示摘要的覆盖范围。 */
export function contextStatusLabels(contexts: ContextStatusRead[] | undefined, names: ReadonlyMap<number, string>): string[] {
  if (!contexts?.length) return ["上下文状态未记录"];
  const phases = { chat: "回答上下文", rewrite: "检索改写上下文", answer: "RAG 回答上下文" };
  return contexts.map((item) => {
    const compressed = item.compressed === null ? "压缩状态未知" : item.compressed ? "已压缩" : "未压缩";
    const tokens = item.estimatedTokens === null ? "输入估计量未知" : `输入估计 ${item.estimatedTokens} Token`;
    const history = item.historyThroughTurn === null ? "历史范围未知" : item.historyThroughTurn === 0 ? "无历史轮次" : `历史截至第 ${item.historyThroughTurn} 轮`;
    return `${names.get(item.modelConfigId) ?? `模型 ${item.modelConfigId}`} · ${phases[item.phase]} · ${compressed} · ${tokens} · ${history}`;
  });
}

File: backend/app/services/multiturn/context_status.py
"""从已保存上下文中投影最小状态，不返回内部消息或摘要正文。"""
from app.models.conversation import ConversationContext
from app.schemas.conversation import ContextStatusRead


def project_context_status(row: ConversationContext) -> list[ContextStatusRead]:
    """兼容普通和 RAG 分阶段快照，非法或缺失元数据保留未知。"""
    snapshot = row.snapshot_json or {}
    phases = snapshot.get("rag_requests")
    records = [(phase, phases[phase]) for phase in ("rewrite", "answer") if phase in phases] if isinstance(phases, dict) else [("chat", snapshot)]
    result = []
    for phase, value in records:
        value = value if isinstance(value, dict) else {}
        count = value.get("estimated_tokens")
        compressed = value.get("compressed")
        result.append(ContextStatusRead(modelConfigId=row.model_config_id, phase=phase,
            compressed=compressed if type(compressed) is bool else None,
            estimatedTokens=count if type(count) is int and count >= 0 else None,
            historyThroughTurn=row.covered_through_turn))
    return result
